File: code-samples/messaging/message-store-export.js
const RC  = require('@ringcentral/sdk').SDK
var fs    = require('fs')
var https = require('https');
require('dotenv').config();

var rcsdk = new RC({
    'server':       process.env.RC_SERVER_URL,
    'clientId':     process.env.RC_CLIENT_ID,
    'clientSecret': process.env.RC_CLIENT_SECRET
});

var platform = rcsdk.platform();

platform.login({ 'jwt':  process.env.RC_JWT })

platform.on(platform.events.loginSuccess, function(e) {
  create_message_store_report()
});

async function create_message_store_report() {
  try {
    var endpoint = "/restapi/v1.0/account/~/message-store-report"
    var params = {
      dateFrom: "2019-03-01T00:00:00.000Z",
      dateTo: "2019-03-31T23:59:59.999Z"
    }
    var resp = await platform.post(endpoint, params)
    var jsonObj = await resp.json()
    get_message_store_report_task(jsonObj.id)
  } catch (e) {
    console.log(e)
  }
}

async function get_message_store_report_task(taskId) {
  console.log("check task creation status ...")
  try {
    var endpoint = "/restapi/v1.0/account/~/message-store-report/" + taskId
    var resp = await platform.get(endpoint)
    var jsonObj = await resp.json()
    if (jsonObj.status == "Completed") {
      get_message_store_report_archive(jsonObj.id)
    } else {
      setTimeout(function() {
        get_message_store_report_task(taskId)
      }, 10000);
    }
  } catch (e) {
    console.log(e)
  }
}

async function get_message_store_report_archive(taskId) {
  console.log("getting report uri ...")
  try {
    var endpoint = `/restapi/v1.0/account/~/message-store-report/${taskId}/archive`
    var resp = await platform.get(endpoint)
    var jsonObj = await resp.json()
    var date = new Date()
    for (var i = 0; i < jsonObj.records.length; i++) {
      var fileName = "message_store_content_" + date.toISOString() + "_" + i + ".zip"
      get_message_store_report_archive_content(jsonObj.records[i].uri, fileName)
    }
  } catch (e) {
    console.log(e)
  }
}

async function get_message_store_report_archive_content(contentUri, fileName){
  var arr = contentUri.split("//")
  var index = arr[1].indexOf('/')
  var domain = arr[1].substring(0, index)
  var path = arr[1].substring(index, arr[1].length)
  var tokenObj = await platform.auth().data()
  var accessToken = tokenObj.access_token
  download(domain, path, accessToken, fileName, function(){
    console.log("Save atttachment to the local machine.")
  })
}

const download = function(domain, path, accessToken, dest, cb) {
  var file = fs.createWriteStream(dest);
  var options = {
          host: domain,
          path: path,
          method: "GET",
          headers: {
            Authorization: `Bearer ${accessToken}`
          }
      }
  const req = https.request(options, res => {
    res.pipe(file);
    file.on('finish', function() {
      file.close(cb);
    });
  })
  req.on('error', error => {
    console.error(error)
  })
  req.end()
}

File: code-samples/messaging/message-store-export.py
#!/usr/bin/env python
from ringcentral import SDK
from datetime import date
from urllib.request import urlopen
import os,sys,time
from dotenv import load_dotenv
load_dotenv()

CLIENTID     = os.environ.get('RC_CLIENT_ID')
CLIENTSECRET = os.environ.get('RC_CLIENT_SECRET')
SERVER       = os.environ.get('RC_SERVER_URL')
JWT          = os.environ.get('RC_JWT')

sdk = SDK( CLIENTID, CLIENTSECRET, SERVER )
platform = sdk.platform()
platform.login( jwt = JWT )

today = date.today()
dateLog = today.strftime("%Y_%m_%d_%H_%M")
ZIPFILE= "message_store_" + dateLog  + ".zip"

def create_message_store_report() :
    endpoint = "/restapi/v1.0/account/~/message-store-report"
    params = {
	"dateFrom": "2019-01-01T00:00:00.000Z",
	"dateTo": "2019-03-31T23:59:59.999Z"
    }
    response = platform.post(endpoint, params)
    json = response.json()
    get_message_store_report_task(json.id)

def get_message_store_report_task(taskId):
    print("check task creation status ...")
    endpoint = "/restapi/v1.0/account/~/message-store-report/" + taskId
    response = platform.get(endpoint)
    json = response.json()
    if json.status == "Completed":
        return get_message_store_report_archive(taskId)
    else:
        time.sleep(2)
        return get_message_store_report_task(taskId)

def get_message_store_report_archive(taskId):
    print("getting report uri ...")
    endpoint = "/restapi/v1.0/account/~/message-store-report/"+ taskId +"/archive"
    jsonObj = platform.get(endpoint).json()
    for i in range( len(jsonObj.records) ):
        get_message_store_report_archive_content( jsonObj.records[i].uri, i )

def get_message_store_report_archive_content(contentUri, i):
    zipFileName = f'{i}-' + ZIPFILE
    print( f'Save report to: {zipFileName}')
    uri = platform.create_url(contentUri, False, None, True);
    fileHandler = urlopen(uri)
    with open(zipFileName, 'wb') as output:
        output.write(fileHandler.read())

create_message_store_report()
sys.exit(0)
